databinding newsapi kotlin

OyaCanli opened the pull request:
Added the kotlin version of NewsApi databinding sample

rajtheinnovator commented on DatabindingWithNewsApiKotlin/app/src/main/java/com/enpassio/databindingwithnewsapikotlin/data/NewsRepository.kt:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    private inner class NewsAsyncTask : AsyncTask<Void, Void, List<Article>>() {
```
I'm not able to re-create the issue always but while testing the app, I found this: 
```
FATAL EXCEPTION: main
    Process: com.enpassio.databindingwithnewsapikotlin, PID: 5400
    java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter list
        at com.enpassio.databindingwithnewsapikotlin.data.NewsRepository$NewsAsyncTask.onPostExecute(NewsRepository.kt)
        at com.enpassio.databindingwithnewsapikotlin.data.NewsRepository$NewsAsyncTask.onPostExecute(NewsRepository.kt:38)
```
If I try starting the app with no internet connection and then turn my wifi on and airplane mode off, the app crashes, **if the internet is not connected(rather just active)**. So we can also check that before making network call, we make sure that it's actually able to do network call. One of the way of doing this can be, making a ping:
```
 private var isInternetConnected = MutableLiveData<Boolean>()
    fun isInternetAvailable(): MutableLiveData<Boolean> {
        networkExecutor.execute {
            var internetConnectionStatus: Boolean
            try {
                val ipAddr = InetAddress.getByName("google.com")
                //You can replace it with your name
                internetConnectionStatus = !ipAddr.equals("")

            } catch (e: Exception) {
                internetConnectionStatus = false
            }

            if (!internetConnectionStatus) {
                val handler = Handler(Looper.getMainLooper())
                handler.post(Runnable {
                    Toast.makeText(
                        MyApplication.context,
                        "Internet not connected. Please try again!",
                        Toast.LENGTH_SHORT
                    ).show()
                })
            }
            isInternetConnected.postValue(internetConnectionStatus)
        }
        return isInternetConnected
    }
```
Ref: @link: https://stackoverflow.com/a/9570292

OyaCanli replied:
@rajtheinnovator Well I can't replicate the issue but I believe you :) Which device/emulator do you use? I had come across this "pinging internet for checking connection" concept before but didn't understand much. Now it makes sense. I'll search a bit about this. 
And putting connectionStatus in a livedata is a great idea. Why did I bother with the callbacks! 
And I'll need to change this in java version as well.

rajtheinnovator replied:
@OyaCanli I used two devices for it. OnePlus3T and a Xiaomi api 27 device. I guess the reason you're not able to replicate the issue is that by the time you make api call, internet gets connected and so it works fine. I tried running it where connectivity was not seamless and sometimes it took few seconds to get connected to internet. 

I think the code would work most of the time but yes, it can create issue if we're looking for super smooth experience, specially in production ready apps. Not critical but maybe worth trying to take a look. The code I shared above is from one of the app I was working on and there I saw its need. It worked well for my use case and I think you can try it too. Just let me know if you any issue in this or Java version :+1:

OyaCanli commented:
@rajtheinnovator I added two commits: 
-I corrected the return type as nullable. I hope this will fix the crash.
-I added an enum for UIState, and put it in an observable field in viewmodel. It facilitates the switch between loading, success and error states
-I moved network check to fragment instead of repository. It was complicating the code with listeners and callbacks for no reason. 
For pinging internet for connection, some of the methods proposed in SO are returning false although there is internet. And some others require using background thread. So it requires some extra work. So may be you try it like this, to see whether that nullable type correction solved the crash. Else, we'll discuss which way to go for pinging internet. 

These two didn't work:
```//Returns false although there is internet
public static Boolean isOnline() {
        try {
            Process p1 = java.lang.Runtime.getRuntime().exec("ping -c 1 www.google.com");
            int returnVal = p1.waitFor();
            return (returnVal == 0);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

//Returns false although there is internet
public static boolean thereIsInternet(){
        boolean internetConnectionStatus;
        try {
            InetAddress ipAddr = InetAddress.getByName("google.com");
            //You can replace it with your name
            internetConnectionStatus = !ipAddr.toString().equals("");
        } catch (Exception e) {
            internetConnectionStatus = false;
        }
        return internetConnectionStatus;
    } ```

OyaCanli commented:
@rajtheinnovator I think I found a way of making a real internet connection check without complicating the process with extra asynctasks and listeners. I tried it on java version. I'll push that one so you can take a look. If it seems better to you, I'll change kotlin version accordingly.
